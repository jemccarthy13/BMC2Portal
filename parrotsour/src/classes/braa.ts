import { drawText } from "../canvas/draw/drawutils"
import { lpad } from "../utils/mathutilities"

/**
 * BRAA is a class that contains a bearing and range
 * as well as formatting and drawing functions for any bearing/range.
 */
export class BRAA {
  bearing = "000"
  bearingNum = 0
  range = 0
  label = ""

  constructor(bearingNum: number, range: number) {
    this.bearingNum = bearingNum
    this.bearing = lpad(this.bearingNum, 3)
    this.range = range
  }

  /**
   * Formatter
   *
   * @returns Formatted bearing and range
   */
  public toString(): string {
    return this.label + " " + this.bearing + "/" + this.range
  }

  /**
   * Draw a bearing and range at the given coordinates.
   *
   * @param ctx Current drawing context
   * @param x X position of draw
   * @param y Y position of draw
   * @param color
   * @param showMeasurements
   */
  draw(
    ctx: CanvasRenderingContext2D,
    x: number,
    y: number,
    color?: string,
    showMeasurements?: boolean
  ): void {
    if (showMeasurements) {
      drawText(ctx, this.bearing + "/" + this.range, x, y, 11, color)
    }
  }
}
