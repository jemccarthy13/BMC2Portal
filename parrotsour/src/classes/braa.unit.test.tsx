import { BRAA } from "./braa"

// eslint-disable-next-line @typescript-eslint/no-var-requires
const canvasSerializer = require("jest-canvas-snapshot-serializer")

expect.addSnapshotSerializer(canvasSerializer)

describe("BRAA Class", () => {
  let canvas: HTMLCanvasElement
  let ctx: CanvasRenderingContext2D

  beforeAll(() => {
    canvas = document.createElement("canvas")
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    ctx = canvas.getContext("2d")!
    canvas.height = 20
    canvas.width = 50
  })

  afterEach(() => {
    ctx.clearRect(0, 0, canvas.width, canvas.height)
  })

  it("constructABraa", () => {
    const br = new BRAA(90, 20)
    expect(br.bearing).toEqual("090")
    expect(br.range).toEqual(20)
    expect(br.bearingNum).toEqual(90)
  })

  it("BraaToString", () => {
    const br = new BRAA(90, 20)
    expect(br.toString()).toEqual("090/20")
  })

  it("drawsCorrectly", () => {
    const br = new BRAA(90, 20)
    br.draw(ctx, 10, 10, "black", true)
    expect(canvas).toMatchSnapshot()
  })

  it("drawsOnlyOnShowMeasurements", () => {
    const br = new BRAA(90, 20)
    br.draw(ctx, 10, 10, "black", false)
    expect(canvas).toMatchSnapshot()
  })
})
