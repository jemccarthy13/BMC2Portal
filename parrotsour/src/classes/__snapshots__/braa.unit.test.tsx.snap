// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BRAA Class drawsCorrectly 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIAAAAUCAYAAADPym6aAAAABmJLR0QA/wD/AP+gvaeTAAABQklEQVRIie3UvytFcRjH8VeS34UJGdxJGZWBiQxiUJRV3VWRP4B/RCkrpQwGs8Gg/BgkXEr4AyQRE8P3q3vcuOfqWuS863SenvP5PM/5fs/zPWT8XYZwhiccYjTm+7Af81toS9FDDtcYxDGecYSBCrxVs4UFtGAZ2zG/hxk0YAIrKXrIYw0XmIzeOZxU4K2aO8Xd7sZljJ9iQ2iOunJ6cRH5kvpNuK/AWzUvqI1xPR5jfIBxYVfn8ZqiJ4xVrqT+mOLXLOetmsdYlPDSDzHuF0biAUuJpt/pc8JCknRiB60p3l/hCh0x7sLpF5pexTn/Tp8XRkvi2Tp6ftjrEzVpggT7mEVjvBdivoAptGMamyn6YezGeBwbWMRNBb1+hRHcCmfgXPhFfuQvhXFYRV2KPnk+bvBWcpXzZmRkZGRkZPwL3gHchmKGFUrUNwAAAABJRU5ErkJggg=="
  height="20"
  width="50"
/>
`;

exports[`BRAA Class drawsOnlyOnShowMeasurements 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIAAAAUCAYAAADPym6aAAAABmJLR0QA/wD/AP+gvaeTAAAAG0lEQVRIie3BMQEAAADCoPVPbQ0PoAAAAIBHAw+0AAGPFGGuAAAAAElFTkSuQmCC"
  height="20"
  width="50"
/>
`;
